### PR TITLE
[Login] Login control

### PR DESCRIPTION
**Pull request by lingma**
## Brief summary of changes

­login is accessible after user login, the current PR will redirect user to the dashboard if that happens.

#### Testing instructions (if applicable)

1. Login then check /login link

#### Link(s) to related issue(s)

@maltheism reported the behavior

### CODE REVIEW

**Comment by lingma** [comment]
OK, I will check, I don't know how to redirect in Loris, so I just show the dashboard to simplify the issue.

**Comment by driusan** [comment]
In the handle function (in this case of the module function), include a "Location" header and appropriate response code in the PSR7 response returned.

**Comment by lingma** [comment]
@driusan This PR is necessary for this release. Because currently if you visit /login, you will ALWAYS see a login page without end if you don't click somewhere else.

File: modules/login/php/module.class.inc [hunk 1]
```diff
@@ -47,4 +49,22 @@ class Module extends \Module
...
+    /**
+     * Check whether a user is logined or not, if yes, redirect to the base URL
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface the outgoing PSR7 response
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        $user = \User::singleton();
```
Comment (driusan) [inline modules/login/php/module.class.inc:62]: The base router includes a "user" attribute on the request which is a (slightly) better place to get the user than a singleton.

File: modules/login/php/module.class.inc [hunk 2]
```diff
@@ -47,4 +49,22 @@ class Module extends \Module
...
+        if (!($user instanceof \LORIS\AnonymousUser)) {
+            return (new \LORIS\Http\Response())
+                ->withStatus(307)
+                ->withHeader("Location", '/');
```
Comment (driusan) [inline modules/login/php/module.class.inc]:
```suggestion
                ->withHeader("Location", \NDB_Factory::singleton()->settings()->getBaseURL());
```

This (untested) will work whether or not LORIS is hosted under the base of the domain.

File: modules/login/php/module.class.inc [hunk 3]
```diff
@@ -47,4 +49,22 @@ class Module extends \Module
...
+     *
+     * @return ResponseInterface the outgoing PSR7 response
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        $user = \User::singleton();
+        if (!($user instanceof \LORIS\AnonymousUser)) {
+            return (new \LORIS\Http\Response())
+                ->withStatus(307)
+                ->withHeader('Location', '/');
```
Comment (driusan) [inline modules/login/php/module.class.inc:66]:
```suggestion
                ->withHeader('Location', \NDB_Factory::singleton()->settings()->getBaseURL());
```

This (untested) will work regardless of whether LORIS is hosted under the base of the domain name.

**Comment by driusan** [comment]
@maltheism can you retest since it's been rewritten since you added the tag?

File: modules/login/php/module.class.inc [hunk 4]
```diff
@@ -47,4 +49,25 @@ class Module extends \Module
     {
         return "Login";
     }
+
+    /**
+     * Check whether a user is logined or not, if yes, redirect to the base URL
```
Comment (driusan) [inline modules/login/php/module.class.inc:54]:
```suggestion
     * Check whether a user is logged in or not, if yes, redirect to the base URL
```

**Pull request merged**